File: src/content/docs/topics/cs/coding-problems/intervals/435-non-overlapping-intervals-approach3.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function eraseOverlapIntervals(intervals: number[][]): number {
    intervals.sort((a, b) => a[0] - b[0]);
    let count = 0;
    let prevEnd = -Infinity;
    for (const [s, e] of intervals) {
        if (s >= prevEnd) {
            prevEnd = e;
        } else {
            count++;
            prevEnd = Math.min(prevEnd, e);
        }
    }
    return count;
}

assert(eraseOverlapIntervals([[1,2],[2,3],[3,4],[1,3]]) === 1);
assert(eraseOverlapIntervals([[1,2],[1,2],[1,2]]) === 2);
assert(eraseOverlapIntervals([[1,2],[2,3]]) === 0);
assert(eraseOverlapIntervals([[1,5]]) === 0);
assert(eraseOverlapIntervals([]) === 0);
assert(eraseOverlapIntervals([[1,100],[2,3],[4,5],[6,7]]) === 1);
console.log("all tests pass");

File: src/content/docs/topics/cs/coding-problems/intervals/435-non-overlapping-intervals-approach3.py
def erase_overlap_intervals(intervals):
    intervals.sort(key=lambda x: x[0])
    count = 0
    prev_end = float('-inf')
    for s, e in intervals:
        if s >= prev_end:
            prev_end = e
        else:
            count += 1
            prev_end = min(prev_end, e)
    return count

assert erase_overlap_intervals([[1,2],[2,3],[3,4],[1,3]]) == 1
assert erase_overlap_intervals([[1,2],[1,2],[1,2]]) == 2
assert erase_overlap_intervals([[1,2],[2,3]]) == 0
assert erase_overlap_intervals([[1,5]]) == 0
assert erase_overlap_intervals([]) == 0
assert erase_overlap_intervals([[1,100],[2,3],[4,5],[6,7]]) == 1
print("all tests pass")
